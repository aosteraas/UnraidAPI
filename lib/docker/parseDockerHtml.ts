import { DockerContainer } from '@models/docker';
import { parse } from 'node-html-parser';

export function parseDockerHtml(rawHtml: string): DockerContainer[] {
  const document = parse(rawHtml);

  const rows = document.querySelectorAll('tr.sortable');

  const containers = rows.map((row) => {
    const name = row.querySelector('.appname').firstChild.innerText;
    const imageUrl = row.querySelector('img.img').getAttribute('src');
    const containerId = row.querySelector('.hand').id;
    const status = row.querySelector('.state').innerText;

    const updatecolumn = row.querySelector('.updatecolumn');
    const useLastChild = updatecolumn.childNodes.length === 3;
    const tag = updatecolumn.childNodes[useLastChild ? 2 : 0].textContent;

    return { name, imageUrl, status, containerId, tag };
  });

  return containers;
}
