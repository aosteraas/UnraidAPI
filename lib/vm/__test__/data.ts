export const vmHtml = `<tr parent-id="0" class="sortable">
  <td class="vm-name" style="width: 220px; padding: 8px">
    <span class="outer"
      ><span id="vm-1d608caf-a589-395f-121a-c6c70f9abac6" class="hand"
        ><img
          src="/plugins/dynamix.vm.manager/templates/images/ubuntu.png"
          class="img" /></span
      ><span class="inner"
        ><a
          href="#"
          onclick='return toggle_id("name-0")'
          title="click for more VM info"
          >Lubuntu</a
        ><br /><i class="fa fa-square stopped red-text"></i
        ><span class="state">stopped</span></span
      ></span
    >
  </td>
  <td></td>
  <td>
    <a class="vcpu-1d608caf-a589-395f-121a-c6c70f9abac6" style="cursor: pointer"
      >2</a
    >
  </td>
  <td>4096M</td>
  <td title="Current physical size: 20G">1 / 20G</td>
  <td>VNC:auto</td>
  <td>
    <input
      class="autostart"
      type="checkbox"
      name="auto_Lubuntu"
      title="Toggle VM autostart"
      uuid="1d608caf-a589-395f-121a-c6c70f9abac6"
    />
  </td>
</tr>
<tr child-id="0" id="name-0" style="display: none">
  <td colspan="8" style="margin: 0; padding: 0">
    <table class="tablesorter domdisk" id="domdisk_table">
      <thead>
        <tr>
          <th><i class="fa fa-hdd-o"></i> <b>Disk devices</b></th>
          <th>Bus</th>
          <th>Capacity</th>
          <th>Allocation</th>
        </tr>
      </thead>
      <tbody id="domdisk_list">
        <tr>
          <td>/mnt/disks/vms/Lubuntu/vdisk1.img</td>
          <td>VirtIO</td>
          <td title="Click to increase Disk Size">
            <form method="get" action="">
              <input type="hidden" name="subaction" value="disk-resize" /><input
                type="hidden"
                name="uuid"
                value="1d608caf-a589-395f-121a-c6c70f9abac6"
              /><input
                type="hidden"
                name="disk"
                value="/mnt/disks/vms/Lubuntu/vdisk1.img"
              /><input type="hidden" name="oldcap" value="20G" /><span
                class="diskresize"
                style="width: 30px"
                ><span class="text"
                  ><a href="#" onclick="return false">20G</a></span
                ><input
                  class="input"
                  type="text"
                  style="width: 46px"
                  name="cap"
                  value="20G"
                  val="diskresize"
                  hidden
              /></span>
            </form>
          </td>
          <td>19G</td>
        </tr>
        <tr>
          <td>/mnt/user/isos/lubuntu-20.10-desktop-amd64.iso</td>
          <td>SATA</td>
          <td>2G</td>
          <td>2G</td>
        </tr>
      </tbody>
    </table>
  </td>
</tr>
<tr parent-id="1" class="sortable">
  <td class="vm-name" style="width: 220px; padding: 8px">
    <span class="outer"
      ><span id="vm-3c15a77b-7917-514e-91e5-3aa32c29f919" class="hand"
        ><img
          src="/plugins/dynamix.vm.manager/templates/images/ubuntu.png"
          class="img" /></span
      ><span class="inner"
        ><a
          href="#"
          onclick='return toggle_id("name-1")'
          title="click for more VM info"
          >Ubuntu</a
        ><br /><i class="fa fa-square stopped red-text"></i
        ><span class="state">stopped</span></span
      ></span
    >
  </td>
  <td></td>
  <td>
    <a class="vcpu-3c15a77b-7917-514e-91e5-3aa32c29f919" style="cursor: pointer"
      >1</a
    >
  </td>
  <td>1024M</td>
  <td title="Current physical size: 2G">1 / 2G</td>
  <td>VNC:auto</td>
  <td>
    <input
      class="autostart"
      type="checkbox"
      name="auto_Ubuntu"
      title="Toggle VM autostart"
      uuid="3c15a77b-7917-514e-91e5-3aa32c29f919"
    />
  </td>
</tr>
<tr child-id="1" id="name-1" style="display: none">
  <td colspan="8" style="margin: 0; padding: 0">
    <table class="tablesorter domdisk" id="domdisk_table">
      <thead>
        <tr>
          <th><i class="fa fa-hdd-o"></i> <b>Disk devices</b></th>
          <th>Bus</th>
          <th>Capacity</th>
          <th>Allocation</th>
        </tr>
      </thead>
      <tbody id="domdisk_list">
        <tr>
          <td>/mnt/disks/vms/Ubuntu/vdisk1.img</td>
          <td>VirtIO</td>
          <td title="Click to increase Disk Size">
            <form method="get" action="">
              <input type="hidden" name="subaction" value="disk-resize" /><input
                type="hidden"
                name="uuid"
                value="3c15a77b-7917-514e-91e5-3aa32c29f919"
              /><input
                type="hidden"
                name="disk"
                value="/mnt/disks/vms/Ubuntu/vdisk1.img"
              /><input type="hidden" name="oldcap" value="2G" /><span
                class="diskresize"
                style="width: 30px"
                ><span class="text"
                  ><a href="#" onclick="return false">2G</a></span
                ><input
                  class="input"
                  type="text"
                  style="width: 46px"
                  name="cap"
                  value="2G"
                  val="diskresize"
                  hidden
              /></span>
            </form>
          </td>
          <td>1M</td>
        </tr>
      </tbody>
    </table>
  </td>
</tr>
addVMContext('Lubuntu','1d608caf-a589-395f-121a-c6c70f9abac6','Ubuntu','shutoff','','');addVMContext('Ubuntu','3c15a77b-7917-514e-91e5-3aa32c29f919','Ubuntu','shutoff','','');var
kvm=[];kvm.push({id:'1d608caf-a589-395f-121a-c6c70f9abac6',state:'shutoff'});kvm.push({id:'3c15a77b-7917-514e-91e5-3aa32c29f919',state:'shutoff'});`;
