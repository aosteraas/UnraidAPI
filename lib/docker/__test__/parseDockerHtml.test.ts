import { parseDockerHtml } from '../parseDockerHtml';
import { dockerHtml } from './data';

test('Processes raw Docker HTML into array of containers', async () => {
  const res = parseDockerHtml(dockerHtml);
  expect(res).toMatchSnapshot();
});

test.todo('Processes raw Docker HTML including images');
