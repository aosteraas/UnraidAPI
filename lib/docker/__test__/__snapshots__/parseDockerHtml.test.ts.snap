// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Processes raw Docker HTML into array of containers 1`] = `
Array [
  Object {
    "id": "9938ae3781a6",
    "imageUrl": "/state/plugins/dynamix.docker.manager/images/AirConnect-icon.png?1625388428",
    "name": "
      ",
    "status": "started",
    "tag": "
      ",
  },
  Object {
    "id": "a84b20a370de",
    "imageUrl": "/state/plugins/dynamix.docker.manager/images/binhex-plex-icon.png?1625388428",
    "name": "
      ",
    "status": "started",
    "tag": "
      ",
  },
  Object {
    "id": "b63993dcef0f",
    "imageUrl": "/state/plugins/dynamix.docker.manager/images/binhex-sickchill-icon.png?1625388428",
    "name": "
      ",
    "status": "started",
    "tag": "
      ",
  },
  Object {
    "id": "32b1384edfd2",
    "imageUrl": "/state/plugins/dynamix.docker.manager/images/HomeAssistantCore-icon.png?1625388428",
    "name": "
      ",
    "status": "started",
    "tag": "
      ",
  },
  Object {
    "id": "d2e69379be47",
    "imageUrl": "/state/plugins/dynamix.docker.manager/images/NginxProxyManager-icon.png?1625388428",
    "name": "
      ",
    "status": "started",
    "tag": "
      ",
  },
  Object {
    "id": "c3fde85eeea4",
    "imageUrl": "/state/plugins/dynamix.docker.manager/images/postgresql-icon.png?1625388428",
    "name": "
      ",
    "status": "stopped",
    "tag": "
      ",
  },
  Object {
    "id": "09db804639b5",
    "imageUrl": "/state/plugins/dynamix.docker.manager/images/qbittorrentvpn-icon.png?1625388428",
    "name": "
      ",
    "status": "started",
    "tag": "
      ",
  },
  Object {
    "id": "8d28bb268dcb",
    "imageUrl": "/state/plugins/dynamix.docker.manager/images/redis-icon.png?1625388428",
    "name": "
      ",
    "status": "stopped",
    "tag": "
      ",
  },
  Object {
    "id": "713c6cc1b2b9",
    "imageUrl": "/state/plugins/dynamix.docker.manager/images/shinobi_pro-icon.png?1625388428",
    "name": "
      ",
    "status": "started",
    "tag": "
      ",
  },
  Object {
    "id": "f84c9a6e1535",
    "imageUrl": "/state/plugins/dynamix.docker.manager/images/TeamCity-icon.png?1625388428",
    "name": "
      ",
    "status": "started",
    "tag": "
      ",
  },
  Object {
    "id": "60f2290f844c",
    "imageUrl": "/state/plugins/dynamix.docker.manager/images/TeamCity-agent-nodejs-icon.png?1625388428",
    "name": "
      ",
    "status": "started",
    "tag": "
      ",
  },
  Object {
    "id": "ef5c77482592",
    "imageUrl": "/state/plugins/dynamix.docker.manager/images/Unraid-API-icon.png?1625388428",
    "name": "
      ",
    "status": "stopped",
    "tag": "
      ",
  },
]
`;
